feature/home-screen

MuhammedWael9991 opened the pull request.

MuhammedWael9991 committed:
add new strings for photo addition, example leak, and category item

MuhammedWael9991 committed:
add RequestTitleContent composable with selection chips and input field

MuhammedWael9991 committed:
refactor CategoryItem composable to use CategoryIconBox for icon rendering and update previews

MuhammedWael9991 committed:
add AddPhotosContent composable for adding photo functionality

MuhammedWael9991 committed:
fix standardize 'Ok' string to 'OK' for consistency

MuhammedWael9991 committed:
fix: enforce single line input for RequestTitleContent

MuhammedWael9991 committed:
Merge remote-tracking branch 'origin/develop' into feature/home-components

# Conflicts:
#	app/src/main/res/values-ar/strings.xml
#	app/src/main/res/values/strings.xml

MuhammedWael9991 committed:
Merge remote-tracking branch 'origin/develop' into feature/home-components

# Conflicts:
#	app/src/main/res/values-ar/strings.xml
#	app/src/main/res/values/strings.xml

MuhammedWael9991 committed:
refactor rename UserRepositoryImp to UserRepositoryImpl and add getStats method

MuhammedWael9991 committed:
add GetStatsUseCase to retrieve user statistics

MuhammedWael9991 committed:
refactor rename UserRemoteDataSourceImp to UserRemoteDataSourceImpl for consistency

MuhammedWael9991 committed:
add GetStatsException for handling user statistics retrieval errors

MuhammedWael9991 committed:
add Stats and StatsDto data classes

MuhammedWael9991 committed:
add GetStatsUseCase to use case module

MuhammedWael9991 committed:
refactor rename UserRemoteDataSourceImp to UserRemoteDataSourceImpl and update package structure; add getStats function

MuhammedWael9991 committed:
refactor rename UserRemoteDataSourceImp to UserRemoteDataSourceImpl and update package structure; add getStats function

MuhammedWael9991 committed:
add getStats function to UserRepository interface

MuhammedWael9991 committed:
add GetStatsException and GetRecentRelatedJobsException to San3aExceptionTree

MuhammedWael9991 committed:
add RequestService, RequestServiceDto, and RequestServiceMapper for service data handling

MuhammedWael9991 committed:
add getRecentRelatedJobs function to UserRemoteDataSource and UserRemoteDataSourceImpl

MuhammedWael9991 committed:
add getRecentRelatedJobs function to UserRepository and UserRepositoryImpl

MuhammedWael9991 committed:
add GetRecentRelatedJobsUseCase to useCaseModule

MuhammedWael9991 committed:
add GetRecentRelatedJobsUseCase for fetching recent related jobs

MuhammedWael9991 committed:
refactor GetRecentRelatedJobsUseCase

MuhammedWael9991 committed:
refactor HomeRepository to remove unused methods and add searchServices function

MuhammedWael9991 committed:
add SearchServicesException for handling search services errors

MuhammedWael9991 committed:
add searchServices function to ServiceRemoteDataSource and its implementation

MuhammedWael9991 committed:
add SearchServiceUseCase for searching services

M-Abdelnabi committed:
Merge branch 'develop' into feature/home-domain

M-Abdelnabi committed:
feat: Add requestedCount to RequestService and RequestServiceDto

This commit introduces a new field, `requestedCount`, to both the `RequestService` domain model and its corresponding DTO, `RequestServiceDto`.

The `toDomain()` and `toDto()` mapper functions have been updated to include this new field.

M-Abdelnabi committed:
feat: Add requestedCount to RequestService and RequestServiceDto

This commit introduces a new field, `requestedCount`, to both the `RequestService` domain model and its corresponding DTO, `RequestServiceDto`.

The `toDomain()` and `toDto()` mapper functions have been updated to include this new field.

M-Abdelnabi committed:
Add requestedCount field to RequestServiceDto

The `requestedCount` field has been added to the `RequestServiceDto` data class. This field represents the number of times a service has been requested.

The `fromJson` companion object function has been updated to parse the `requestedCount` from the input data map. If the `requestedCount` is not present or is not a Long, it defaults to 0.

M-Abdelnabi committed:
Add requestedServiceDto parameter to requestService method

The requestService method in the ServiceRemoteDataSource interface now accepts a RequestServiceDto object as a parameter. This allows for more detailed information about the service being requested.

Additionally, a new method, getMostRequestedServices, has been added to retrieve a list of the most requested services for a given user.

M-Abdelnabi committed:
feat: Implement requestService and getMostRequestedServices in ServiceRemoteDataSource

This commit implements the `requestService` and `getMostRequestedServices` functions in `ServiceRemoteDataSourceImpl`.

The `requestService` function now handles incrementing the request count for existing services or creating a new service request entry in Firestore.

The `getMostRequestedServices` function retrieves a list of services ordered by their request count for a given user from Firestore.

A new constant `SERVICE_REQUESTS_COLLECTION` has been added to define the Firestore collection for service requests.

M-Abdelnabi committed:
feat: Implement requestService and getMostRequestedServices in ServiceRemoteDataSource

This commit implements the `requestService` and `getMostRequestedServices` functions in `ServiceRemoteDataSourceImpl`.

The `requestService` function now handles incrementing the request count for existing services or creating a new service request entry in Firestore.

The `getMostRequestedServices` function retrieves a list of services ordered by their request count for a given user from Firestore.

A new constant `SERVICE_REQUESTS_COLLECTION` has been added to define the Firestore collection for service requests.

M-Abdelnabi committed:
Feat: Add MostRequestServiceMapper

This commit introduces a new mapper, `MostRequestServiceMapper`, which is responsible for converting `RequestServiceDto` objects to `MostRequestedServices` domain entities.

The `toMostRequestedServices()` extension function within this mapper facilitates the transformation by mapping the respective fields from the DTO to the domain entity.

M-Abdelnabi committed:
feat: Add GetMostRequestedServicesException

This commit introduces a new custom exception, `GetMostRequestedServicesException`, to handle errors specifically related to fetching the most requested services.

M-Abdelnabi committed:
feat: Add requestService and getMostRequestedServices functions to HomeRepository

This commit introduces two new functions to the `HomeRepository` interface:

- `requestService`: Allows requesting a service.
- `getMostRequestedServices`: Retrieves the most requested services for a given user.

M-Abdelnabi committed:
feat: Implement service request and most requested services functionality

This commit introduces the ability for users to request services and retrieve a list of the most requested services.

Changes include:
- Added `requestService` function to `HomeRepositoryImpl` to handle service requests.
- Added `getMostRequestedServices` function to `HomeRepositoryImpl` to fetch the most requested services for a given user.
- Implemented corresponding data mapping functions (`toDto`, `toMostRequestedServices`).
- Defined new exceptions (`GetMostRequestedServicesException`) for handling errors related to these new functionalities.

M-Abdelnabi committed:
Refactor: Change `suspend fun invoke()` to `operator fun invoke()` in GetAllServicesUseCase

The `invoke` function in `GetAllServicesUseCase` has been changed from a `suspend fun` to an `operator fun`. This allows the use case to be called directly as a function, simplifying its usage.

M-Abdelnabi committed:
feat: Implement service request and most requested services functionality

This commit introduces the ability for users to request services and retrieve a list of the most requested services.

Changes include:
- Added `requestService` function to `HomeRepositoryImpl` to handle service requests.
- Added `getMostRequestedServices` function to `HomeRepositoryImpl` to fetch the most requested services for a given user.
- Implemented corresponding data mapping functions (`toDto`, `toMostRequestedServices`).
- Defined new exceptions (`GetMostRequestedServicesException`) for handling errors related to these new functionalities.

M-Abdelnabi committed:
feat: Add RequestServiceUseCase

This commit introduces the `RequestServiceUseCase` class.

The `RequestServiceUseCase` class is responsible for handling the business logic related to requesting a service. It takes a `HomeRepository` as a dependency in its constructor.

The class has a single `invoke` operator function that takes a `RequestService` object as a parameter and calls the `requestService` method on the `HomeRepository`.

M-Abdelnabi committed:
feat: Add RequestServiceUseCase

This commit introduces the `RequestServiceUseCase` class.

The `RequestServiceUseCase` class is responsible for handling the business logic related to requesting a service. It takes a `HomeRepository` as a dependency in its constructor.

The class has a single `invoke` operator function that takes a `RequestService` object as a parameter and calls the `requestService` method on the `HomeRepository`.

M-Abdelnabi committed:
Merge branch 'develop' into feature/home-components

# Conflicts:
#	app/src/main/res/values-ar/strings.xml
#	app/src/main/res/values/strings.xml

M-Abdelnabi committed:
feat: Implement RequestCardForCraftsMan Composable

This commit introduces the `RequestCardForCraftsMan` Composable function.

This component is designed to display request details for craftsmen, including:
- Title
- Type
- Offers
- Description
- Location

It also includes a preview function `RequestCardForCraftsManPreview` for easy visualization and testing.

M-Abdelnabi committed:
Remove unused string resource

The string "tap_to_add_photos" was removed from both English and Arabic string resource files as it was no longer being used in the application.

MuhammedWael9991 committed:
refactor GetMostRequestedServicesUseCase to remove userId parameter and return List<Service>

MuhammedWael9991 committed:
Add GetAvailableJobsUseCase for retrieving available jobs

MuhammedWael9991 committed:
Update HomeRepository to include getAvailableJobs and refactor getMostRequestedServices

MuhammedWael9991 committed:
refactor update getMostRequestedServices to remove userId parameter and add getAvailableJobs

MuhammedWael9991 committed:
Add GetAvailableJobsException for handling available jobs retrieval errors

MuhammedWael9991 committed:
Add CustomerHomeInteractionListener and CustomerHomeViewModel for customer home screen functionality

MuhammedWael9991 committed:
Add CraftsmanHomeViewModel and CraftsmanInteractionListener for craftsman home screen functionality

MuhammedWael9991 committed:
Add HomeUiState data classes for managing home screen state

MuhammedWael9991 committed:
Add GetAvailableJobsUseCase and update view model bindings for home screens

MuhammedWael9991 committed:
Refactor HomeScreen to use CustomerHomeViewModel and CraftsmanHomeViewModel

MuhammedWael9991 committed:
Add imports for CraftsmanHomeViewModel and CustomerHomeViewModel in viewModelModule

MuhammedWael9991 committed:
Add SearchBar composable for enhanced user input

M-Abdelnabi committed:
feat: add location to AppBar

This commit introduces the ability to display a location in the AppBar component.

- Added an optional `location` parameter to the `AppBar` composable.
- When `location` is provided, it's displayed below the title with a location icon and a dropdown arrow.
- Adjusted padding for better visual alignment.
- Added a new preview `AppBarPrev2` to demonstrate the AppBar with a location.

MuhammedWael9991 committed:
Add SearchBar composable for enhanced user input

MuhammedWael9991 committed:
Merge remote-tracking branch 'origin/feature/home-components' into feature/home-components

# Conflicts:
#	app/src/main/java/com/paris_2/san3a/presentation/shared/components/SearchBar.kt

M-Abdelnabi committed:
update AppBar component

- Remove `actionIcon` composable parameter.
- Add `onActionIconClick` callback parameter.
- Display a default notification icon if `onActionIconClick` is provided.
- Remove `ExperimentalMaterial3Api` annotation as it's no longer needed.
- Update preview functions to reflect changes.

MuhammedWael9991 committed:
Add RequestDescriptionContent composable for user input

M-Abdelnabi committed:
Refactor: Update HomeRepository to use MostRequestedServices entity

The `getMostRequestedServices` function in `HomeRepository` now returns a Flow of `List<MostRequestedServices>` instead of `List<Service>`.

M-Abdelnabi committed:
Refactor: Adapt getMostRequestedServices to use MostRequestedServices entity

The `getMostRequestedServices` method in `HomeRepositoryImpl` has been updated.

It now returns a `Flow<List<MostRequestedServices>>` instead of `Flow<List<Service>>`.
This change involves mapping the DTO to the new `MostRequestedServices` entity using `toMostRequestedServices()`.

M-Abdelnabi committed:
Refactor: Change ServiceDto to RequestServiceDto in getMostRequestedServices

The return type of the `getMostRequestedServices` function in `ServiceRemoteDataSource.kt` has been changed from `Flow<List<ServiceDto>>` to `Flow<List<RequestServiceDto>>`.

M-Abdelnabi committed:
Refactor: Change ServiceDto to RequestServiceDto in getMostRequestedServices

The return type of the `getMostRequestedServices` function in `ServiceRemoteDataSource.kt` has been changed from `Flow<List<ServiceDto>>` to `Flow<List<RequestServiceDto>>`.

M-Abdelnabi committed:
Merge remote-tracking branch 'origin/feature/home-components' into feature/home-components

M-Abdelnabi committed:
Add `toMostRequestedServices` mapper

This commit introduces the `toMostRequestedServices` mapper to the `HomeRepositoryImpl.kt` file. This mapper is used to transform service data into a format suitable for displaying the most requested services.

M-Abdelnabi committed:
Refactor: Remove userId from MostRequestServiceMapper

This commit removes the `userId` field from the `toDomain` mapping function in `MostRequestServiceMapper`.

M-Abdelnabi committed:
Refactor: Remove userId from MostRequestServiceMapper

This commit removes the `userId` field from the `toDomain` mapping function in `MostRequestServiceMapper`.

M-Abdelnabi committed:
Refactor: Rename Service entity to MostRequestedServices

The `Service` entity has been renamed to `MostRequestedServices` to better reflect its purpose. This change also updates the `GetMostRequestedServicesUseCase` to use the new entity name.

M-Abdelnabi committed:
Merge branch 'feature/home-domain' into feature/home-screen

M-Abdelnabi committed:
Merge remote-tracking branch 'origin/feature/home-components' into feature/home-screen

MuhammedWael9991 committed:
Add AdCard composable for displaying advertisement content

MuhammedWael9991 committed:
Add new strings for user greetings and job statistics

MuhammedWael9991 committed:
Add new strings for user greetings and job statistics

MuhammedWael9991 committed:
Add new strings for user greetings and job statistics

MuhammedWael9991 committed:
Add vector drawable resources for star, suitcase, and wallet icons

MuhammedWael9991 committed:
Refactor CraftsmanHomeViewModel to use CraftsmanHomeState

MuhammedWael9991 committed:
Rename HomeUiState to CraftsmanHomeUiState and update data structure

MuhammedWael9991 committed:
Add CraftsmanHomeScreen composable and related UI components

M-Abdelnabi committed:
Refactor: Change visibility of `loadUserData` and `loadStats` to private

The methods `loadUserData` and `loadStats` in `CraftsmanHomeViewModel` are now private, restricting their access within the class.

M-Abdelnabi committed:
Refactor: Make load methods private in CustomerHomeViewModel

The `loadServices`, `loadMostRequestedServices`, and `loadUserData` methods in `CustomerHomeViewModel` are now private as they are only used internally within the ViewModel.

Additionally, the ViewModel now uses `CustomerHomeUiState` directly instead of accessing it through `HomeState`.

M-Abdelnabi committed:
Delete HomeScreen.kt

The file `app/src/main/java/com/paris_2/san3a/presentation/screen/HomeScreen.kt` has been removed. This file previously contained a composable function `HomeScreen` that displayed a centered text "Home Screen".

M-Abdelnabi committed:
Refactor: Delete search icon

The search icon drawable resource `ic_search.xml` has been removed.

M-Abdelnabi committed:
Refactor: Add modifier parameter to SearchBar composable

The `SearchBar` composable function now accepts an optional `modifier` parameter, allowing for more flexible layout customization. This parameter is applied to the underlying `BasicTextField`.

M-Abdelnabi committed:
feat: Add GetMostRequestedServicesUseCase to Koin module

The `GetMostRequestedServicesUseCase` has been added to the `useCaseModule` using Koin's `factoryOf` DSL. This makes the use case available for dependency injection throughout the application.

M-Abdelnabi committed:
Define `CustomerUIState` and `CustomerUiState`

This commit introduces two data classes:
- `CustomerHomeUiState`: Represents the overall state of the customer's home screen, including loading status, error messages, and the nested `CustomerUiState`.
- `CustomerUiState`: Holds specific UI data for the customer's home screen, such as the current user's name, location, a list of most requested services, and a list of available services.

M-Abdelnabi committed:
feat: Implement MostRequestedServices component

This commit introduces the `MostRequestedServices` composable function, which displays a horizontally scrollable list of "most requested" services.

Key changes:
- Creates a `Column` to hold the section title and the `LazyRow` of services.
- Displays the title "Most Requested" using `Theme.textStyle.title.small`.
- Uses `LazyRow` to efficiently display a list of `MostRequestedServices`.
- Each service is rendered as a `CategoryItem` with a static icon (ic_washing_machine_bold).
- Applies padding and spacing for visual appeal.

M-Abdelnabi committed:
feat: Add customer home screen and most requested services component

This commit introduces the `CustomerHomeScreen` and the `MostRequestedServices` component.

The `CustomerHomeScreen` displays:
- An app bar with a greeting and location.
- A search bar.
- A section for "Most Requested Services".
- A list of available services under "Find What You Need".
- An "AdCard" to encourage users to become craftsmen.

The `MostRequestedServices` component is a horizontal `LazyRow` that displays `CategoryItem` for each most requested service.

M-Abdelnabi committed:
Refactor: Update requested service count logic

This commit refactors the `requestService` function to correctly update the `requestedCount` of a service.

Previously, if a service request already existed, only the `requestedCount` field was updated. Now, the entire `RequestServiceDto` object is updated using `setDoc` to ensure all fields are correctly persisted. This also addresses a potential issue where `getDoc` might throw an exception if the document doesn't exist, by wrapping it in a try-catch block.

Additionally, the constant `NUMBER_OF_REQUESTS_FIELD` was renamed to `requestedCount` to match the field name in `RequestServiceDto`.

M-Abdelnabi committed:
feat: Add GetMostRequestedServicesUseCase and RequestServiceUseCase

This commit introduces two new use cases:

- `GetMostRequestedServicesUseCase`: Retrieves the most requested services.
- `RequestServiceUseCase`: Handles service requests.

These use cases are now available through dependency injection.

M-Abdelnabi committed:
Refactor: Rename "title" to "name" in ServiceDto

The `title` field in the `ServiceDto` class and its corresponding JSON mapping has been renamed to `name` for better clarity and consistency.

MuhammedWael9991 committed:
Add resource mapping functions for service icons and colors

M-Abdelnabi committed:
Merge remote-tracking branch 'origin/feature/home-screen' into feature/home-screen

MuhammedWael9991 committed:
Add RequestBottomSheetContent composable for handling requests

M-Abdelnabi committed:
Merge remote-tracking branch 'origin/feature/home-screen' into feature/home-screen

MuhammedWael9991 committed:
feat: Update CraftsmanHomeScreen and add search icon vector

MuhammedWael9991 committed:
Add ProgressBar composable for animated progress indication

MuhammedWael9991 committed:
Integrate ProgressBar into RequestBottomSheetContent for step indication

M-Abdelnabi committed:
feat: Add CustomerBottomSheetService composable for service request flow

M-Abdelnabi committed:
Merge remote-tracking branch 'origin/feature/home-screen' into feature/home-screen

M-Abdelnabi committed:
refactor: Update RequestBottomSheetContent to use subTitle instead of description

M-Abdelnabi committed:
feat: Add step indicators to CustomerBottomSheet for improved navigation

MuhammedWael9991 committed:
Add onExitClick callback to CustomerBottomSheet for improved navigation

MuhammedWael9991 committed:
Add onDismissBottomSheet callback to CustomerHomeInteractionListener and implement state update in ViewModel

MuhammedWael9991 committed:
feat: Implement bottom sheet service interaction and state management in CustomerHomeScreen

M-Abdelnabi committed:
feat: Update navigation to use CustomerHomeScreen instead of HomeScreen

MuhammedWael9991 committed:
Add button activation check to back navigation in RequestBottomSheetContent

M-Abdelnabi committed:
feat: Enhance RequestBottomSheetContent layout with background and padding adjustments

M-Abdelnabi committed:
feat: Enhance RequestBottomSheetContent layout with background and padding adjustments

M-Abdelnabi committed:
feat: Update padding in CustomerHomeScreen button for improved layout consistency

MuhammedWael9991 committed:
Add button activation logic based on input fields in CustomerBottomSheet

MuhammedWael9991 committed:
Merge remote-tracking branch 'origin/feature/home-screen' into feature/home-screen

M-Abdelnabi committed:
Fix: Enable click functionality for disabled buttons

This commit addresses an issue where disabled buttons in the RequestBottomSheetContent were still clickable. The fix ensures that the onClickBack and onButtonClick actions are only triggered when the respective buttons (buttonIsActive) are enabled.

M-Abdelnabi committed:
Fix: Correct step navigation in customer bottom sheet

The `onButtonClick` for the problem description step was incorrectly navigating to `SELECT_LOCATION` instead of `IMAGE_UPLOAD`.

The `onClickBack` for the problem description step was incorrectly navigating to `IMAGE_UPLOAD` instead of `SELECT_SERVICE`.

This commit corrects these navigation paths. Additionally, the unused `activeButton` derived state has been removed.

M-Abdelnabi committed:
Fix: Correct step navigation in customer bottom sheet

The `onButtonClick` for the problem description step was incorrectly navigating to `SELECT_LOCATION` instead of `IMAGE_UPLOAD`.

The `onClickBack` for the problem description step was incorrectly navigating to `IMAGE_UPLOAD` instead of `SELECT_SERVICE`.

This commit corrects these navigation paths. Additionally, the unused `activeButton` derived state has been removed.

M-Abdelnabi committed:
Fix: Remove horizontal padding from RequestTitleContent

The horizontal padding on the `RequestTitleContent` component was causing layout issues. This commit removes the unnecessary padding to resolve the problem.

M-Abdelnabi committed:
Remove unnecessary padding in RequestDescriptionContent

The horizontal padding of 16.dp has been removed from the `RequestDescriptionContent` composable as it was deemed unnecessary. This simplifies the layout and potentially improves visual consistency.

M-Abdelnabi committed:
Fix: Change RequestTitleContent text style to medium

The text style for the `RequestTitleContent` composable has been updated from `regular` to `medium` to enhance readability and visual hierarchy.

M-Abdelnabi committed:
Feat: Update request service mapper

Add locationDetails and image to RequestServiceDto and RequestService.

M-Abdelnabi committed:
feat: Add image and locationDetails fields to RequestService

This commit adds the `image` (a list of integers representing image resources) and `locationDetails` (a string) fields to the `RequestService` data class.

M-Abdelnabi committed:
Fix: Enable request acceptance button always

The button to accept a request in the craftsman's view of the bottom sheet was previously disabled until the user interacted with the price and duration input fields. This change ensures the button is always enabled, allowing the craftsman to accept the request with default values if desired.

M-Abdelnabi committed:
Add RequestServiceUiState and mappers

This commit introduces the `RequestServiceUiState` data class to represent the UI state for requesting a service.

It also includes two extension functions:
- `RequestService.toRequestServiceUiState()`: Converts a `RequestService` domain entity to a `RequestServiceUiState`.
- `RequestServiceUiState.toRequestService()`: Converts a `RequestServiceUiState` back to a `RequestService` domain entity, initializing `id`, `relatedJob`, and `offers` with default empty values.

M-Abdelnabi committed:
Fix: Implement `createRequest` function in `CustomerHomeViewModel`

This commit implements the `createRequest` function in `CustomerHomeViewModel` to handle service requests.

The function attempts to execute the `requestServicesUseCase` and updates the UI state accordingly:
- On success, it sets `bottomSheetState` to `true`.
- On error, it updates `errorMessage` with the error message and logs the error.

M-Abdelnabi committed:
Add padding to AppBar and handle service request creation

The `CustomerHomeScreen` now includes padding at the top of the `AppBar`.
Additionally, it initiates a service request creation when a `RequestServiceUiState` is available from the bottom sheet. Logging has been added to track the `RequestServiceUiState`.

M-Abdelnabi committed:
Add padding to AppBar and handle service request creation

The `CustomerHomeScreen` now includes padding at the top of the `AppBar`.
Additionally, it initiates a service request creation when a `RequestServiceUiState` is available from the bottom sheet. Logging has been added to track the `RequestServiceUiState`.

M-Abdelnabi committed:
Feat: Implement Customer Bottom Sheet for Request Creation

This commit introduces the `CustomerBottomSheetService` composable, which guides users through a multi-step process for creating service requests.

The bottom sheet includes steps for:
- Selecting a service
- Providing a description
- Specifying a location
- Uploading images

The `RequestServiceUiState` is updated with the user's input at each step, and the final request details are logged upon completion. Navigation between steps and exiting the bottom sheet are also handled.

M-Abdelnabi committed:
Feat: Update RequestService model to include serviceType and customerComplain

M-Abdelnabi committed:
Feat: Update RequestServiceDto to include serviceType, customerComplain, locationDetails, and image

M-Abdelnabi committed:
Feat: Update MostRequestedSection to display serviceType instead of title

M-Abdelnabi committed:
Feat: Update MostRequestedServiceMapper to use serviceType instead of title

M-Abdelnabi committed:
Feat: Update RequestServiceUiState and MostRequestedServices to use serviceType and customerComplain

M-Abdelnabi committed:
Feat: Update CustomerHomeScreen to use serviceType instead of title for MostRequestedServices

M-Abdelnabi committed:
Refactor: Update RequestServiceUiState and data source path

This commit updates the `RequestServiceUiState` by renaming the `title` field to `serviceType` and adding a new `customerComplain` field.

Additionally, the `ServiceRemoteDataSourceImpl` has been modified to use the `serviceType` field instead of `id` when constructing the document path for Firestore.

M-Abdelnabi committed:
Conditionally display "Most Requested Services"

The "Most Requested Services" section in the `CustomerHomeScreen` will now only be displayed if the `mostRequestedServices` list in `state.customerUiState` is not empty.

M-Abdelnabi committed:
Refactor: Update search bar hint text

The hint text in the search bar on the `CustomerHomeScreen` has been changed from "Search for service" to "Search...". The initial value of the search bar is now an empty string.

ahmedNaser7 committed:
add images picker to use with request services and handle mapping it

ahmedNaser7 committed:
update most recent section item to make a description take one max line

M-Abdelnabi committed:
Refactor: Remove unnecessary success state update in createRequest method

M-Abdelnabi committed:
Fix: Update condition to display Most Requested Services when the list is not empty

M-Abdelnabi committed:
Fix: Correct serviceType field mapping in RequestServiceDto

The `RequestServiceDto.fromJson` factory method was previously attempting to map the "title" field from the incoming data to the `serviceType` property. This has been corrected to map the "serviceType" field, ensuring accurate data deserialization.

M-Abdelnabi committed:
Merge remote-tracking branch 'origin/feature/home-screen' into feature/home-screen

# Conflicts:
#	app/src/main/java/com/paris_2/san3a/data/mapper/RequestServiceMapper.kt

MuhammedWael9991 committed:
Add AddPhotos component for photo management in bottom sheet

MuhammedWael9991 committed:
Add permissions for external storage access in AndroidManifest

MuhammedWael9991 committed:
Refactor job type references

MuhammedWael9991 committed:
Update requestService method to return a String and simplify Firestore interaction

MuhammedWael9991 committed:
Refactor RequestServiceMapper to include title and streamline image handling

MuhammedWael9991 committed:
Refactor RequestService and RequestServiceDto to include title and streamline image handling

MuhammedWael9991 committed:
Refactor MostRequestServiceMapper to remove unused requestedCount field

MuhammedWael9991 committed:
Refactor RequestServiceUiState to include title and remove unused fields

MuhammedWael9991 committed:
Refactor CustomerHomeViewModel to remove unnecessary blank lines

MuhammedWael9991 committed:
Refactor CustomerBottomSheet to enhance image handling and streamline photo addition

MuhammedWael9991 committed:
Merge remote-tracking branch 'origin/develop' into feature/home-screen

# Conflicts:
#	app/src/main/java/com/paris_2/san3a/data/repository/UserRepositoryImpl.kt
#	app/src/main/java/com/paris_2/san3a/data/source/remote/user/UserRemoteDataSource.kt
#	app/src/main/java/com/paris_2/san3a/data/source/remote/user/UserRemoteDataSourceImpl.kt
#	app/src/main/java/com/paris_2/san3a/di/dataModule.kt
#	app/src/main/java/com/paris_2/san3a/di/repositoryModule.kt
#	app/src/main/java/com/paris_2/san3a/di/useCaseModule.kt
#	app/src/main/java/com/paris_2/san3a/di/viewModelModule.kt
#	app/src/main/java/com/paris_2/san3a/domain/San3aExceptionTree.kt
#	app/src/main/java/com/paris_2/san3a/domain/repository/UserRepository.kt
#	app/src/main/java/com/paris_2/san3a/presentation/navigation/NavGraph.kt
#	app/src/main/res/values-ar/strings.xml
#	app/src/main/res/values/strings.xml

MuhammedWael9991 committed:
Merge remote-tracking branch 'origin/develop' into feature/home-screen

# Conflicts:
#	app/src/main/java/com/paris_2/san3a/data/repository/UserRepositoryImpl.kt
#	app/src/main/java/com/paris_2/san3a/data/source/remote/user/UserRemoteDataSource.kt
#	app/src/main/java/com/paris_2/san3a/data/source/remote/user/UserRemoteDataSourceImpl.kt
#	app/src/main/java/com/paris_2/san3a/di/useCaseModule.kt
#	app/src/main/java/com/paris_2/san3a/domain/repository/UserRepository.kt
#	app/src/main/java/com/paris_2/san3a/presentation/navigation/NavGraph.kt
#	app/src/main/res/values-ar/strings.xml
#	app/src/main/res/values/strings.xml

MuhammedWael9991 committed:
fix: improve error handling by providing default message in CraftsmanHomeViewModel

MuhammedWael9991 committed:
refactor: enhance CustomerBottomSheet with location selection and service suggestions

MuhammedWael9991 committed:
feat: add location handling and user data loading in CustomerHomeViewModel

MuhammedWael9991 committed:
feat: update HomeRepository to handle service requests and add update functionality

MuhammedWael9991 committed:
feat: add SuggestionMapper to provide service-specific suggestions

MuhammedWael9991 committed:
feat: add UpdateNumOfRequestsUseCase to use case module

MuhammedWael9991 committed:
feat: update UserRemoteDataSourceImpl to fetch stats from craftsmen collection

MuhammedWael9991 committed:
feat: add updateNumOfRequestService method and exception handling for service requests

MuhammedWael9991 committed:
feat: add UpdateNumOfRequestsUseCase to invoke updateNumOfRequestService

MuhammedWael9991 committed:
feat: remove commented-out getResourceId function from ServiceResourceMapper

MuhammedWael9991 committed:
feat: update MostRequestedServices component to handle service clicks and support Arabic localization

MuhammedWael9991 committed:
feat: update RequestBottomSheetContent to conditionally handle button clicks

MuhammedWael9991 committed:
feat: update RequestService.toDto to provide a default value for imageUrls

MuhammedWael9991 committed:
feat: update GetMostRequestedServicesUseCase to return a list of Service entities

MuhammedWael9991 committed:
feat: update CustomerHomeInteractionListener to include serviceId in createRequest and add updateNumOfRequests method

MuhammedWael9991 committed:
feat: enhance CustomerHomeScreen to support service selection and localization

MuhammedWael9991 committed:
feat: update CraftsmanHomeViewModel to load specific stats and recent jobs

ahmedNaser7 commented on app/src/main/java/com/paris_2/san3a/presentation/shared/components/AppBar.kt:
```diff
@@ -19,12 +21,12 @@ import com.paris_2.san3a.R
 @Composable
 fun AppBar(
     modifier: Modifier = Modifier,
-    actionIcon: @Composable () -> Unit = {},
```
why you delete action icon composable ?

M-Abdelnabi replied:
i solve it

M-Abdelnabi committed:
Refactor: Remove HomeScreen composable

The `HomeScreen` composable, which previously displayed a simple "Home Screen" text, has been removed. Its functionality will likely be integrated into more specific screen composables.

M-Abdelnabi committed:
solve some conflicts

M-Abdelnabi committed:
Fix: Correct Arabic translation for "tap to add photos"

The Arabic translation for "tap to add photos" has been corrected from "اضغط لأضافة صورة" to "اضغط لإضافة صورة" in `app/src/main/res/values-ar/strings.xml`.

M-Abdelnabi committed:
refactor call of function

M-Abdelnabi committed:
Add localizations for customer home screen and remove logging

This commit introduces Arabic localizations for several strings on the customer home screen, including:
- "Find What You Need"
- "Got a skill? Start earning"
- "Create your craftsman account and get job requests."
- "Become a Craftsman"

Additionally, a log statement for `RequestServiceUiState` in `CustomerHomeScreen` has been removed.

M-Abdelnabi committed:
Refactor: Use constants for localization keys in CustomerHomeScreen

This commit replaces hardcoded string keys for localized service titles and descriptions with constants (`ARABIC_NAME`, `ENGLISH_NAME`, `ARABIC_DESCRIPTION`, `ENGLISH_DESCRIPTION`).

The `@Preview` composable `PreviewHomeScreen` has also been removed.

M-Abdelnabi committed:
Add enum class for BottomSheetStep to manage UI states

M-Abdelnabi committed:
Add bottomSheetStep to CustomerUIState for managing BottomSheet navigation

M-Abdelnabi committed:
Add bottomSheetServiceTitle and bottomSheetServiceId to CustomerUIState

M-Abdelnabi committed:
Add bottomSheetIconRes and bottomSheetDescription to CustomerUIState

M-Abdelnabi committed:
Add bottomSheetImages to CustomerUIState for enhanced image support

M-Abdelnabi committed:
Add bottomSheetSelectedSuggestion to CustomerUIState for suggestion management

M-Abdelnabi committed:
Add bottomSheetGovernments and bottomSheetCities to CustomerUIState for location management

M-Abdelnabi committed:
Add bottomSheetSelectedGovernment and bottomSheetSelectedCity to CustomerUIState for enhanced location selection

M-Abdelnabi committed:
Add bottomSheetAddressDetails and visibility flags to CustomerUIState for improved UI management

M-Abdelnabi committed:
Add missing comma in CustomerUIState for improved data structure integrity

M-Abdelnabi committed:
Add initBottomSheet method to CustomerHomeInteractionListener for bottom sheet initialization

M-Abdelnabi committed:
Add initBottomSheet method to CustomerHomeViewModel for bottom sheet state management

M-Abdelnabi committed:
Add updateBottomSheetStep method to CustomerHomeInteractionListener for bottom sheet step management

M-Abdelnabi committed:
Add updateBottomSheetStep method to CustomerHomeViewModel for bottom sheet step updates

M-Abdelnabi committed:
Add nextBottomSheetStep method to CustomerHomeInteractionListener for bottom sheet navigation

M-Abdelnabi committed:
Add nextBottomSheetStep method to CustomerHomeViewModel for bottom sheet navigation

M-Abdelnabi committed:
Add previousBottomSheetStep method to CustomerHomeInteractionListener for bottom sheet navigation

M-Abdelnabi committed:
Add previousBottomSheetStep method to CustomerHomeViewModel for bottom sheet navigation

M-Abdelnabi committed:
Add setBottomSheetServiceTitle method to CustomerHomeInteractionListener for title management

M-Abdelnabi committed:
Add setBottomSheetServiceTitle method to CustomerHomeViewModel for title management

M-Abdelnabi committed:
Add setBottomSheetDescription method to CustomerHomeViewModel and CustomerHomeInteractionListener for description management

M-Abdelnabi committed:
Add setBottomSheetSelectedSuggestion method to CustomerHomeViewModel and CustomerHomeInteractionListener for suggestion management

M-Abdelnabi committed:
Add addBottomSheetImages method to CustomerHomeViewModel and CustomerHomeInteractionListener for image management

M-Abdelnabi committed:
Add deleteBottomSheetImageAt method to CustomerHomeViewModel and CustomerHomeInteractionListener for image removal

M-Abdelnabi committed:
Implement address details update in customer home

This commit introduces functionality to update the address details within the customer home screen's bottom sheet.

The `CustomerHomeViewModel` now includes a `setBottomSheetAddressDetails` function to update the `bottomSheetAddressDetails` in the screen state.

The `CustomerHomeInteractionListener` interface has been updated to include the `setBottomSheetAddressDetails` method, enabling communication of address changes from the UI to the ViewModel.

M-Abdelnabi committed:
Refactor: Add governorate selection to bottom sheet

This commit introduces the ability for users to select their governorate in the service request bottom sheet.

The `CustomerHomeInteractionListener` interface now includes a `setBottomSheetSelectedGovernment` function, and the `CustomerHomeViewModel` implements this function to update the `bottomSheetSelectedGovernment` field in the UI state.

M-Abdelnabi committed:
Feat: Add city selection to bottom sheet

This commit introduces the ability for users to select a city in the bottom sheet.

The `CustomerHomeViewModel` now includes a `setBottomSheetSelectedCity` function to update the `bottomSheetSelectedCity` in the screen state.

The `CustomerHomeInteractionListener` interface has been updated to include the `setBottomSheetSelectedCity` function.

M-Abdelnabi committed:
Feat: Add government sheet visibility control

This commit introduces the ability to show or hide a "government sheet" in the customer home screen.

- A `showGovernmentSheet` function has been added to `CustomerHomeInteractionListener` and implemented in `CustomerHomeViewModel`.
- Calling this function updates the `isGovernmentSheetVisible` flag in the screen state, controlling the visibility of the sheet.

M-Abdelnabi committed:
Add city selection to bottom sheet

This commit introduces the ability to select a city within the bottom sheet for service requests.

The `CustomerHomeViewModel` and `CustomerHomeInteractionListener` have been updated to include:
- `showCitySheet` function: Controls the visibility of the city selection sheet.
- `isCitySheetVisible` state: Tracks whether the city selection sheet is currently visible.

M-Abdelnabi committed:
Feat: Implement bottom sheet state reset

This commit introduces the `resetBottomSheetState` function in `CustomerHomeViewModel` and its corresponding interface method in `CustomerHomeInteractionListener`.

This function resets all bottom sheet related states to their initial values, including:
- `bottomSheetStep` to `SELECT_SERVICE`
- Clearing text fields like `bottomSheetServiceTitle`, `bottomSheetDescription`, `bottomSheetAddressDetails`
- Resetting selections for `bottomSheetServiceId`, `bottomSheetIconRes`, `bottomSheetSelectedSuggestion`, `bottomSheetSelectedGovernment`, `bottomSheetSelectedCity`
- Clearing `bottomSheetImages`
- Hiding `isGovernmentSheetVisible` and `isCitySheetVisible`

M-Abdelnabi committed:
Refactor: Update CustomerHomeScreen with Multi-Step Bottom Sheet

This commit refactors the `CustomerHomeScreen` to implement a multi-step bottom sheet for creating service requests. The `CustomerBottomSheetService` has been replaced with a more generic `BottomSheet` component and a series of `RequestBottomSheetContent` screens for each step of the process.

The bottom sheet now guides users through the following steps:
1.  **Select Service:** Users can select a service and provide a title for their request.
2.  **Problem Description:** Users describe the issue in detail.
3.  **Select Location:** Users specify their location, including government, city, and address details.
4.  **Image Upload:** Users can upload images related to their request.

The `CustomerHomeViewModel` has been updated to manage the state of the bottom sheet, including the current step, selected service, description, location, and images. New actions have been added to handle navigation between steps, updating the bottom sheet state, and creating the service request.

Key changes include:
-   Introduction of `BottomSheetStep` enum to manage the current step in the bottom sheet.
-   Updated `CustomerHomeUiState` to include new properties for managing the bottom sheet state.
-   Implementation of `RequestTitleContent`, `RequestDescriptionContent`, `LocationContent`, and `AddPhotosContent` composables for each step of the bottom sheet.
-   Refactoring of `CustomerHomeScreenContent` to use the new bottom sheet implementation.
-   Integration of `rememberLauncherForActivityResult` for handling image selection from the gallery.
-   Updated `MostRequestedServices` to initialize the bottom sheet with the selected service title and ID.

M-Abdelnabi committed:
Refactor: Remove CustomerBottomSheet composable

The `CustomerBottomSheetService` composable has been removed. This component was responsible for handling the multi-step process of creating a service request, including selecting a service, providing a description, specifying a location, and uploading images.

M-Abdelnabi committed:
Refactor: Improve CustomerHomeViewModel logic

This commit introduces the following changes to `CustomerHomeViewModel`:

- The `getCities` function is now private, limiting its accessibility.
- The `initBottomSheet` function is called within `onServiceClick` to initialize the bottom sheet with service details.
- The `resetBottomSheetState` function is called within `onExitClick` to clear the bottom sheet state when the user exits.

M-Abdelnabi committed:
Refactor: Rename `bottomSheetServiceTitle` to `bottomSheetSubtitle`

The `bottomSheetServiceTitle` property in `CustomerUIState` has been renamed to `bottomSheetSubtitle` to better reflect its purpose.

Correspondingly, the `setBottomSheetServiceTitle` function in `CustomerHomeInteractionListener` and `CustomerHomeViewModel` has been renamed to `setBottomSheetServiceSubTitle`. This change improves the clarity and consistency of the codebase.

M-Abdelnabi committed:
Refactor: Update bottom sheet title handling

The `CustomerHomeScreen` now uses `bottomSheetSubtitle` and `setBottomSheetServiceSubTitle` for managing the title in the request creation bottom sheet. This replaces the previous usage of `bottomSheetServiceTitle` and `setBottomSheetServiceTitle`.

The `onChipClick` action now updates the subtitle with the selected suggestion.

M-Abdelnabi committed:
Refactor: Remove redundant update of `bottomSheetServiceTitle`

The `bottomSheetServiceTitle` was being unnecessarily updated with the `suggestion` value in the `updateBottomSheetSelectedSuggestion` function. Since `bottomSheetServiceTitle` is already initialized and represents the general service, this redundant update has been removed. The title will now remain the initial service title even when a suggestion is selected.

M-Abdelnabi committed:
Refactor: Clear bottom sheet subtitle on step reset

The `bottomSheetSubtitle` in `CustomerHomeViewModel` is now cleared when resetting the bottom sheet step to `SELECT_SERVICE`. This ensures that any previously set subtitle from other steps is removed.

M-Abdelnabi committed:
Refactor: Update RequestTitleContent text color

The text color of the `RequestTitleContent` composable has been updated to use `Theme.colors.shade.primary`. This change ensures consistency with the app's color scheme.

M-Abdelnabi committed:
Fix: Set text color for request description input

The text color for the request description input field in `RequestDescriptionContent` is now explicitly set to `Theme.colors.shade.primary`. This ensures consistent text color with other input fields.

M-Abdelnabi committed:
Refactor: Update `AppTextField` disabled state colors

The `AppTextField` component has been updated to use specific theme colors for its disabled state.

This includes setting distinct colors for:
- Disabled text
- Disabled border
- Disabled placeholder
- Disabled leading icon
- Disabled trailing icon
- Disabled label

M-Abdelnabi committed:
Fix: Enable clickable modifier for location text field

The `clickable` modifier for the location `AppTextField` in `LocationContent.kt` was previously not explicitly enabled. This commit sets the `enabled` parameter of the `clickable` modifier to `true` to ensure it always responds to clicks.

M-Abdelnabi committed:
Fix: Show city sheet after government selection

The `CustomerHomeScreen` now automatically displays the city selection sheet after the user selects a government in the bottom sheet. Previously, both government and city sheets were shown simultaneously when "Get Location" was clicked.

M-Abdelnabi committed:
Refactor: Adjust bottom padding in RequestTitleContent

The bottom padding for the `LazyRow` within the `RequestTitleContent` composable has been removed. The padding was previously set to `24.dp` and is now `0.dp`. The top padding remains at `12.dp`.

M-Abdelnabi committed:
Fix: Update subtitles for address and image upload steps

The subtitles for the address details and image upload steps in the customer home screen's bottom sheet have been updated.

- The address details step subtitle is now "Where are you from?".
- The image upload step subtitle is now "Add some photos".

M-Abdelnabi committed:
Refactor: Enhance AppBar with flexible action icon

The `AppBar` composable has been updated to accept a generic `actionIcon` composable lambda.

This change allows for greater flexibility in defining the action icon displayed in the `AppBar`, replacing the previous hardcoded icon. The `onActionIconClick` callback remains to handle interactions with this custom action icon.

M-Abdelnabi committed:
Refactor: Separate BottomSheetUiState from CustomerUiState

This commit refactors the `CustomerUIState` by extracting the bottom sheet related states into a new data class `BottomSheetUiState`.

- `CustomerUIState` now includes a `bottomSheetUiState` property of type `BottomSheetUiState`.
- The following properties were moved from `CustomerUIState` to `BottomSheetUiState`:
    - `bottomSheetState`
    - `bottomSheetStep`
    - `bottomSheetServiceTitle`
    - `bottomSheetDescription`
    - `bottomSheetLocation`
    - `bottomSheetImage`
    - `isGovernmentSheetVisible`
    - `isCitySheetVisible`

This separation improves the organization and clarity of the UI state classes.

M-Abdelnabi committed:
Refactor: Encapsulate bottom sheet state in `BottomSheetUiState`

This commit introduces the `BottomSheetUiState` data class to encapsulate all UI state related to the bottom sheet in `CustomerHomeViewModel`.

All functions that previously modified individual bottom sheet properties now update the `bottomSheetUiState` object within the main `CustomerHomeUiState`. This improves code organization and readability by grouping related state variables.

M-Abdelnabi committed:
Refactor: Group bottom sheet UI state in `CustomerHomeScreen`

This commit refactors the `CustomerHomeScreen` to group all bottom sheet related UI state variables into a new data class `BottomSheetUiState` within the main `CustomerHomeUiState`.

This improves code organization and readability by encapsulating the bottom sheet's state, making it easier to manage and understand the different UI states related to the bottom sheet functionality.

All references to individual bottom sheet state variables have been updated to access them through the new `bottomSheetUiState` object.

M-Abdelnabi committed:
Translate bottom sheet UI strings

This commit translates the following strings used in the customer home screen's bottom sheet to Arabic and adds them to `strings.xml`:

- "Select a service"
- "Describe the problem in detail"
- "Describe your problem"
- "Where are you from?"
- "Add some photos"
- "Create Request"

The English versions of these strings were also added to `strings.xml` for consistency. The `CustomerHomeScreen.kt` was updated to use these string resources.

M-Abdelnabi committed:
feat(ui): Enhance RequestBottomSheetContent with optional text and image support

This commit updates the `RequestBottomSheetContent` composable to:

- Display an optional text next to the subtitle. This is useful for indicating optional fields or providing additional context.
- Introduce a `containImg` parameter, which is currently unused but paves the way for future image display within the bottom sheet.

The preview has been updated to showcase the new optional text feature.

M-Abdelnabi committed:
Fix: Add "optional" text to image upload step

The image upload step in the service request creation process now displays "(optional)" next to the "Add some photos" subtitle. This clarifies that providing images is not mandatory.

M-Abdelnabi committed:
Translate bottom sheet strings to Arabic

This commit adds Arabic translations for the following strings used in the customer bottom sheet:

- "Describe the problem in detail"
- "Where are you from?"
- "Add some photos"
- "Create Request"
- "Select a service"
- "Describe your problem"

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/data/repository/HomeRepositoryImpl.kt:
```diff
@@ -11,31 +19,65 @@ import kotlinx.coroutines.flow.map
 
-    override suspend fun getRecentRelatedJobs() {
-        TODO("Not yet implemented")
+    override suspend fun updateNumOfRequestService(serviceId: String) {
+        return try {
+            serviceRemoteDataSource.updateNumOfRequestService(serviceId)
+        }catch (e: Exception) {
+            throw UpdateNumOfRequestsException()
+        }
     }
```
use safe call in suspend functions

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/data/repository/HomeRepositoryImpl.kt:
```diff
@@ -11,31 +19,65 @@ import kotlinx.coroutines.flow.map
+            serviceRemoteDataSource.requestService(requestedService.toDto(imageUrls))
+        } catch (e: Exception) {
+            throw e
+        }
```
use safe call in suspend functions

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/data/source/remote/service/ServiceRemoteDataSourceImpl.kt:
```diff
@@ -16,11 +21,58 @@ class ServiceRemoteDataSourceImpl(
+        return fireStoreService.streamCollection(
+            path = SERVICES_COLLECTION,
+            fromJson = ServiceDto::fromJson,
+            limit = null,
```
remove this line it's not required

Joseph-Sameh-0 replied:
limit = null,

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/data/source/remote/user/dto/RequestServiceDto.kt:
```diff
@@ -0,0 +1,45 @@
+                "userId" to userId
+            )
+            return map
+        }
```
move the toJson outside the companion object scope

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/screen/home/craftsman/CraftsmanHomeViewModel.kt:
```diff
@@ -0,0 +1,98 @@
+
+    init {
+        loadUserData()
+        loadStats("aIqNtF3KQ84r5x6jc9O1")
```
why the id is hard coded ?

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/screen/home/craftsman/CraftsmanHomeViewModel.kt:
```diff
@@ -0,0 +1,98 @@
+    init {
+        loadUserData()
+        loadStats("aIqNtF3KQ84r5x6jc9O1")
+        loadRecentRelatedJobs("Plumbing")
```
why hard coded values?

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/screen/home/craftsman/CraftsmanHomeViewModel.kt:
```diff
@@ -0,0 +1,98 @@
+                updateState(
+                    screenState.value.copy(
+                        craftsmanHomeUiState = screenState.value.craftsmanHomeUiState.copy(
+                            stats = it ?: throw IllegalStateException("Stats cannot be null")
```
why you throw here?
if the state is null the state part should disappear from the screen

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/screen/home/craftsman/CraftsmanHomeViewModel.kt:
```diff
@@ -0,0 +1,98 @@
+    }
+    fun loadRecentRelatedJobs(relatedJob: String){
+        tryToExecute(
+            execute = { getRecentRelatedJobsUseCase(relatedJob).first() },
```
don't use .first with flow 
on success make .collect
and update the state inside it to make it Realtime

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/screen/home/craftsman/CraftsmanHomeViewModel.kt:
```diff
@@ -0,0 +1,98 @@
+    }
+    fun loadAvailableJobs(){
+        tryToExecute(
+            execute = { getAvailableJobsUseCase().first() },
```
again .first() ?!!

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/screen/home/customer/component/MostRequestedSection.kt:
```diff
@@ -0,0 +1,66 @@
+        modifier = modifier
+    ) {
+        Text(
+            text = "Most Requested",
```
localize the strings

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/screen/home/customer/CustomerHomeScreen.kt:
```diff
@@ -0,0 +1,301 @@
+            item {
+                SearchBar(
+                    value = "",
+                    onValueChange = { /*TODO*/ },
```
?

M-Abdelnabi replied:
search feature would be implemented in a separate PR

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/screen/home/customer/CustomerHomeScreen.kt:
```diff
@@ -0,0 +1,301 @@
+                SearchBar(
+                    value = "",
+                    onValueChange = { /*TODO*/ },
+                    hint = "Search...",
```
localize the strings

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/screen/home/customer/CustomerHomeViewModel.kt:
```diff
@@ -0,0 +1,394 @@
+                    )
+                )
+            )
+        }
```
use try to execute

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/screen/home/customer/CustomerHomeViewModel.kt:
```diff
@@ -0,0 +1,394 @@
+                )
+            )
+        }
+    }
```
use try to execute

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/screen/home/customer/CustomerHomeViewModel.kt:
```diff
@@ -0,0 +1,394 @@
+
+    private fun loadServices() {
+        tryToExecute(
+            execute = { getAllServicesUseCase().first() },
```
don't use .first() use .collect with flow in the on success scope

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/screen/home/customer/CustomerHomeViewModel.kt:
```diff
@@ -0,0 +1,394 @@
+
+    private fun loadMostRequestedServices() {
+        tryToExecute(
+            execute = { getMostRequestedServicesUseCase().first() },
```
don't use .first() use .collect with flow in the on success scope

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/screen/home/customer/CustomerHomeViewModel.kt:
```diff
@@ -0,0 +1,394 @@
+        navigate(Destinations.Notifications)
+    }
+
+    override fun onSearch(query: String) {}
```
?

M-Abdelnabi replied:
search feature doesn't implemented yet

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/screen/home/utils/ServiceResourceMapper.kt:
```diff
@@ -0,0 +1,56 @@
+import com.paris_2.san3a.R
+import com.paris_2.san3a.presentation.shared.designSystem.theme.Theme
+
+fun getResource(id: String): Int {
+    return when(id){
+        "00FoToEQx8OAVK8fprck" -> R.drawable.ic_paint_roller_bold
+        "0AyHLfbkAS6aFnHeZL6Z" -> R.drawable.ic_plug_circle_bold
+        "N1XbpJwF27ByQxGJzGcW" -> R.drawable.ic_washing_machine_bold
+        "VMuPvjaMBy7xZQB2om6l" -> R.drawable.ic_conditioner_bold
+        "ig1I47LLgLUHg96d8ZwB" -> R.drawable.ic_leaf_bold
+        "keLsnqeGp8RKpZwWm7J0" -> R.drawable.ic_waterdrops_bold
+        else -> R.drawable.ic_sledgehammer_bold
+    }
+}
```
what is this ?!!!

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/screen/home/utils/ServiceResourceMapper.kt:
```diff
@@ -0,0 +1,56 @@
+    }
+}
+
+@Composable
+fun getResourceColors(id: String): Color {
+    return when(id){
+        "00FoToEQx8OAVK8fprck" -> Theme.colors.additional.secondary.yellow  //painting
+        "0AyHLfbkAS6aFnHeZL6Z" -> Theme.colors.additional.secondary.turquoise // Electrical
+        "N1XbpJwF27ByQxGJzGcW" -> Theme.colors.additional.secondary.purple // Washing
+        "VMuPvjaMBy7xZQB2om6l" -> Theme.colors.additional.secondary.red // Conditioner
+        "ig1I47LLgLUHg96d8ZwB" -> Theme.colors.additional.secondary.green // Gardening
+        "keLsnqeGp8RKpZwWm7J0" -> Theme.colors.additional.secondary.blue // Plumbing
+        else -> Theme.colors.additional.secondary.blue
+    }
+}
```
Again ?!
Don't depend on the data source id

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/screen/home/utils/ServiceResourceMapper.kt:
```diff
@@ -0,0 +1,56 @@
+    }
+}
+
+@Composable
+fun getResourceTint(id: String): Color {
+    return when(id){
+        "00FoToEQx8OAVK8fprck" -> Theme.colors.additional.primary.yellow
+        "0AyHLfbkAS6aFnHeZL6Z" -> Theme.colors.additional.primary.turquoise
+        "N1XbpJwF27ByQxGJzGcW" -> Theme.colors.additional.primary.purple
+        "VMuPvjaMBy7xZQB2om6l" -> Theme.colors.additional.primary.red
+        "ig1I47LLgLUHg96d8ZwB" -> Theme.colors.additional.primary.green
+        "keLsnqeGp8RKpZwWm7J0" -> Theme.colors.additional.primary.blue
+        else -> Theme.colors.additional.secondary.blue
+    }
+}
```
Again ?!
Don't depend on the id

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/screen/home/utils/ServiceResourceMapper.kt:
```diff
@@ -0,0 +1,56 @@
+//        "Landscaping" -> "ig1I47LLgLUHg96d8ZwB"
+//        "Plumping" -> "keLsnqeGp8RKpZwWm7J0"
+//        else -> ""
+//    }
```
remove this

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/shared/components/AddPhotos.kt:
```diff
@@ -0,0 +1,106 @@
+            ){
+                Icon(
+                    painter = painterResource(R.drawable.ic_plus_outline),
+                    contentDescription = "Add photo",
```
localize the strings

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/shared/components/AppBar.kt:
```diff
@@ -19,12 +21,13 @@ import com.paris_2.san3a.R
+    onActionIconClick: (() -> Unit)? = null,
     actionIcon: @Composable () -> Unit = {},
     title: String,
+    location: String? = null,
```
why the app bar take location ?!!

Joseph-Sameh-0 replied:
It should take Composable for reuse not location

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/shared/components/AppBar.kt:
```diff
@@ -19,12 +21,13 @@ import com.paris_2.san3a.R
 fun AppBar(
     modifier: Modifier = Modifier,
+    onActionIconClick: (() -> Unit)? = null,
     actionIcon: @Composable () -> Unit = {},
```
why you need onActionIconClick while the actionIcon is Composable  ?!

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/shared/components/RequestTitleContent.kt:
```diff
@@ -0,0 +1,152 @@
+            .clip(RoundedCornerShape(16.dp))
+            .border(
+                width = 1.dp,
+                color = if (isSelected) Theme.colors.brand.secondary else Theme.colors.shade.quinary,
```
animate the color

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/shared/components/RequestTitleContent.kt:
```diff
@@ -0,0 +1,152 @@
+                shape = RoundedCornerShape(16.dp)
+            )
+            .background(
+                if (isSelected) Theme.colors.brand.tertiary else Theme.colors.shade.quinary
```
animate the color

M-Abdelnabi committed:
Refactor: Preserve existing bottom sheet state when updating

This commit modifies the `CustomerHomeViewModel` to preserve the existing `BottomSheetUiState` when updating specific properties like `bottomSheetStep`, `bottomSheetSubtitle`, `bottomSheetDescription`, etc.

Previously, updating any of these properties would inadvertently reset other bottom sheet states to their default values. This change ensures that only the intended property is modified, while the rest of the `BottomSheetUiState` remains unchanged. This is achieved by using `screenState.value.bottomSheetUiState.copy(...)` instead of creating a new `BottomSheetUiState` instance.

M-Abdelnabi committed:
Refactor: Use `collect` to observe services flow in `CustomerHomeViewModel`

The `loadServices` function in `CustomerHomeViewModel` has been updated to use the `collect` terminal operator on the Flow returned by `getAllServicesUseCase`.

Previously, `first()` was used, which only collected the first emission. Now, `collect` is used to continuously observe the Flow, ensuring that the UI state is updated whenever new service data is emitted.

M-Abdelnabi committed:
Refactor: Update `loadMostRequestedServices` to collect from Flow

The `loadMostRequestedServices` function in `CustomerHomeViewModel` has been refactored to collect from the Flow returned by `getMostRequestedServicesUseCase` instead of taking the first emitted value.

This change ensures that the UI will reactively update if the most requested services data changes over time.

M-Abdelnabi committed:
Refactor: Localize service title in `CustomerHomeViewModel`

The `onServiceClick` function in `CustomerHomeViewModel` has been updated to display the service title in the user's device language (Arabic or English).

Previously, it defaulted to the first available title. Now, it checks the device's locale and selects the appropriate localized title ("arabicName" or "englishName") from the `selectedService.title` map.

M-Abdelnabi committed:
Refactor: Remove unused `SearchServiceUseCase` from `CustomerHomeViewModel`

This commit removes the unused `SearchServiceUseCase` and its associated imports (`kotlinx.coroutines.flow.collect`, `kotlinx.coroutines.flow.first`, `kotlin.invoke`, `kotlin.text.get`) from the `CustomerHomeViewModel`.

This cleanup improves code maintainability by removing dead code.

M-Abdelnabi committed:
Refactor: Remove unused `getResourceId` function

This commit removes the `getResourceId` function from `ServiceResourceMapper.kt` as it is no longer used.

Joseph-Sameh-0 committed:
refactor: enhance error handling for service requests with custom exception

Joseph-Sameh-0 committed:
Merge branch 'develop' into feature/home-screen

Joseph-Sameh-0 committed:
update authorization token in network module

Joseph-Sameh-0 committed:
add navigation for Craftsman and Customer graphs, enhance CraftsmanHomeScreen layout

Joseph-Sameh-0 committed:
update navigation labels based on user type and introduce LocalUserType state management

Joseph-Sameh-0 committed:
refactor: rename LocalUserType to LocalAccountType and update related navigation logic

Joseph-Sameh-0 committed:
refactor: improve loadAvailableJobs function and clean up code formatting

Joseph-Sameh-0 committed:
update search hint in CustomerHomeScreen and add localized strings

Joseph-Sameh-0 committed:
add localized string for "Most Requested" in MostRequestedSection

Joseph-Sameh-0 committed:
refactor: streamline government and city selection logic in CustomerHomeScreen

Joseph-Sameh-0 committed:
refactor: remove unnecessary limit parameter in ServiceRemoteDataSourceImpl

Joseph-Sameh-0 committed:
refactor: update button activation logic and streamline RequestServiceDto serialization

Joseph-Sameh-0 committed:
refactor: enhance AppBar component with leading icon and streamline title display
